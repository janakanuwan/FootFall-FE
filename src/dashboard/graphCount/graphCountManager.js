import dateTimeUtil from '../../app/utils/dateTimeUtil';
import entriesUtil from '../../app/utils/entriesUtil';
import { GraphData, List } from 'Models';

/**
 *
 * @param dateRange object with 'fromData', 'fromDateMax', 'toDate', 'toDateMin' properties
 * @param type 'from' or 'to'
 * @param date in 'YYYY-MM-DD' format
 * @returns {boolean} true if valid
 */
const isValidDateRange = (dateRange, type, date) => {
  if (type === 'from') {
    return dateTimeUtil.isSameOrAfter(dateRange.toDate, date);
  } else if (type === 'to') {
    return dateTimeUtil.isSameOrAfter(date, dateRange.fromDate);
  }
  return false;
};

/**
 * NOTE: Use {@link deviceCountManager.isValidDateRange} before using this
 *
 * @param dateRange {object} with 'fromData', 'fromDateMax', 'toDate', 'toDateMin' properties
 * @param type 'from' or 'to'
 * @param date in 'YYYY-MM-DD' format
 * @returns {Array} with {param1: value1, param2: value2} format
 */
const updatedDateRangeValues = (dateRange, type, date) => {
  if (type === 'from') {
    return [{ fromDate: date }, { toDateMin: date }];
  } else if (type === 'to') {
    return [{ toDate: date }, { fromDateMax: date }];
  }
  return [];
};

/**
 * NOTE: This highly depends on the {@link GraphDisplayOptions}
 *
 * @param fromTime {number} inclusive
 * @param toTime {number} exclusive
 * @param optionType {string} in {@link GraphDisplayOptions}
 * @return {string} formatted time ban range name
 */
const formatTimeBinRangeName = (fromTime, toTime, optionType) => {
  switch (optionType) {
    case 'hourly':
      return dateTimeUtil.formatDateTime(toTime, 'h:mm A');
    case 'day':
      return dateTimeUtil.formatDateTime(toTime - 1, 'MMM-DD');
    case 'week':
      return `${dateTimeUtil.formatDateTime(fromTime, 'MMM-DD')}:${dateTimeUtil.formatDateTime(toTime - 1, 'MMM-DD')}`;
    case 'month':
      return dateTimeUtil.formatDateTime(toTime - 1, 'MMM YYYY');
    default:
      return dateTimeUtil.formatDateTime(toTime, 'h:mm A');
  }
};

/**
 * NOTE: This highly depends on the {@link GraphDisplayOptions} and the {@link GraphData}
 *
 * @param entries {List} Entry with 'time' and 'locationId' properties
 * @param location {Location} with 'id' property
 * @param dateRange {object} with 'fromData' and 'toDate' properties
 * @param optionType {string} in {@link GraphDisplayOptions}
 * @return {List} the resulting {@link GraphData} list
 */
const computedGraphData = (entries, location, dateRange, optionType = 'hourly') => {
  const graphData = [];

  if (entries.size > 0 && location) {
    const fromTime = dateTimeUtil.millisStartOfDay(dateRange.fromDate);
    const toTime = dateTimeUtil.millisEndOfDay(dateRange.toDate);
    const filteredByRange = entriesUtil.filteredEntries(entries, location, fromTime, toTime);
    const timeBinRange = dateTimeUtil.timeBinRange(fromTime, toTime, optionType);

    for (let i = 0; i < timeBinRange.length - 1; i += 1) {
      const filteredByTime = entriesUtil.filteredEntriesByTime(
        filteredByRange, timeBinRange[i], timeBinRange[i + 1]
      );
      // console.log('Time:', dateTimeUtil.formatDateTime(timeBinRange[i+1]-1));
      graphData.push(GraphData({
        NAME: formatTimeBinRangeName(timeBinRange[i], timeBinRange[i + 1], optionType),
        IN: entriesUtil.sumEntry(filteredByTime),
        OUT: entriesUtil.sumExit(filteredByTime),
        PRESENCE: entriesUtil.netEntry(filteredByTime),
      }));
    }
  }
  return List(GraphData)(graphData);
};


export default {
  isValidDateRange,
  updatedDateRangeValues,
  computedGraphData,
};
