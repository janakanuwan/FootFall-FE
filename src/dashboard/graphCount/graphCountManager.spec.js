import graphCountManager from './graphCountManager';
import { List, Record } from 'Models';

describe('graphCountManager', () => {
  [
    {
      dateRange: { fromDate: '2018-04-04', toDate: '2018-04-04' }, type: 'from', date: '2018-04-04', expected: true,
    },
    {
      dateRange: { fromDate: '2018-04-04', toDate: '2018-04-05' }, type: 'from', date: '2018-04-04', expected: true,
    },
    {
      dateRange: { fromDate: '2018-04-02', toDate: '2018-04-03' }, type: 'from', date: '2018-04-04', expected: false,
    },
    {
      dateRange: { fromDate: '2018-04-04', toDate: '2018-04-04' }, type: 'to', date: '2018-04-04', expected: true,
    },
    {
      dateRange: { fromDate: '2018-04-03', toDate: '2018-04-03' }, type: 'to', date: '2018-04-04', expected: true,
    },
    {
      dateRange: { fromDate: '2018-04-05', toDate: '2018-04-05' }, type: 'to', date: '2018-04-04', expected: false,
    },
  ].forEach(({
    dateRange, type, date, expected,
  }) => {
    it(`should send correct result for 'isValidDateRange' ({${dateRange.fromDate}, ${dateRange.toDate}}, ${type}, ${date})`, () => {
      expect(graphCountManager.isValidDateRange(dateRange, type, date)).toEqual(expected);
    });
  });

  [
    {
      dateRange: {
        fromDate: '2018-04-04', toDate: '2018-04-04', fromDateMax: '2018-04-04', toDateMin: '2018-04-04',
      },
      type: 'from',
      date: '2018-04-04',
      expected: [{ fromDate: '2018-04-04' }, { toDateMin: '2018-04-04' }],
    },
    {
      dateRange: {
        fromDate: '2018-04-03', toDate: '2018-04-04', fromDateMax: '2018-04-04', toDateMin: '2018-04-03',
      },
      type: 'from',
      date: '2018-04-04',
      expected: [{ fromDate: '2018-04-04' }, { toDateMin: '2018-04-04' }],
    },
    {
      dateRange: {
        fromDate: '2018-04-04', toDate: '2018-04-04', fromDateMax: '2018-04-04', toDateMin: '2018-04-04',
      },
      type: 'to',
      date: '2018-04-04',
      expected: [{ toDate: '2018-04-04' }, { fromDateMax: '2018-04-04' }],
    },
    {
      dateRange: {
        fromDate: '2018-04-04', toDate: '2018-04-04', fromDateMax: '2018-04-04', toDateMin: '2018-04-04',
      },
      type: 'to',
      date: '2018-04-05',
      expected: [{ toDate: '2018-04-05' }, { fromDateMax: '2018-04-05' }],
    },
  ].forEach(({
    dateRange, type, date, expected,
  }, index) => {
    it(`should send correct result for 'updatedDateRangeValues' (index:${index})`, () => {
      expect(graphCountManager.updatedDateRangeValues(dateRange, type, date)).toEqual(expected);
    });
  });

  const TestEntry = Record({ entry: Number, exit: Number, locationId: Number, time: Number });

  [
    {
      entries: [
        { entry: 1, exit: 0, locationId: 1, time: 1525116600000 },
        { entry: 2, exit: 1, locationId: 2, time: 1525120200000 },
        { entry: 3, exit: 1, locationId: 1, time: 1525120200000 },
        { entry: 2, exit: 3, locationId: 1, time: 1525123800000 },
        { entry: 5, exit: 6, locationId: 1, time: 1525127400000 },
        { entry: 4, exit: 2, locationId: 1, time: 1525131000000 },
        { entry: 3, exit: 0, locationId: 1, time: 1525134600000 },
        { entry: 0, exit: 1, locationId: 1, time: 1525138200000 },
      ],
      location: { id: 1 },
      dateRange: { fromDate: '2018-05-01', toDate: '2018-05-01' },
      optionType: 'hourly'
    },
    {
      entries: [
        { entry: 1, exit: 0, locationId: 1, time: 1525116600000 },
        { entry: 2, exit: 1, locationId: 2, time: 1525120200000 },
        { entry: 3, exit: 1, locationId: 1, time: 1525120200000 },
        { entry: 2, exit: 3, locationId: 1, time: 1525123800000 },
        { entry: 5, exit: 6, locationId: 1, time: 1525127400000 },
        { entry: 4, exit: 2, locationId: 1, time: 1525131000000 },
        { entry: 3, exit: 0, locationId: 1, time: 1525134600000 },
        { entry: 0, exit: 1, locationId: 1, time: 1525138200000 },
      ],
      location: { id: 1 },
      dateRange: { fromDate: '2018-05-01', toDate: '2018-05-01' },
      optionType: 'hourly'
    },
    {
      entries: [
        { entry: 1, exit: 0, locationId: 1, time: 1525113000000 },
        { entry: 2, exit: 1, locationId: 1, time: 1525120200000 },
        { entry: 1, exit: 2, locationId: 1, time: 1525123800000 },

        { entry: 2, exit: 1, locationId: 1, time: 1525199400000 },
        { entry: 3, exit: 1, locationId: 1, time: 1525285800000 },
        { entry: 2, exit: 3, locationId: 1, time: 1525372200000 },
        { entry: 6, exit: 5, locationId: 1, time: 1525458600000 },
      ],
      location: { id: 1 },
      dateRange: { fromDate: '2018-05-01', toDate: '2018-05-04' },
      optionType: 'day'
    },
    {
      entries: [
        { entry: 1, exit: 0, locationId: 1, time: 1525113000000 },
        { entry: 2, exit: 1, locationId: 1, time: 1525120200000 },
        { entry: 1, exit: 2, locationId: 1, time: 1525123800000 },
        { entry: 2, exit: 1, locationId: 1, time: 1525199400000 },

        { entry: 3, exit: 1, locationId: 1, time: 1525717800000 },
        { entry: 2, exit: 3, locationId: 1, time: 1526322600000 },
        { entry: 6, exit: 5, locationId: 1, time: 1526581800000 },
      ],
      location: { id: 1 },
      dateRange: { fromDate: '2018-05-01', toDate: '2018-05-18' },
      optionType: 'week'
    },
    {
      entries: [
        { entry: 1, exit: 0, locationId: 1, time: 1525113000000 },
        { entry: 2, exit: 1, locationId: 1, time: 1525120200000 },
        { entry: 1, exit: 2, locationId: 1, time: 1525123800000 },
        { entry: 2, exit: 1, locationId: 1, time: 1525199400000 },

        { entry: 3, exit: 1, locationId: 1, time: 1527791400000 },
        { entry: 2, exit: 3, locationId: 1, time: 1530383400000 },
        { entry: 6, exit: 5, locationId: 1, time: 1533061800000 },
        { entry: 4, exit: 2, locationId: 1, time: 1533666600000 },
      ],
      location: { id: 1 },
      dateRange: { fromDate: '2018-05-01', toDate: '2018-09-20' },
      optionType: 'month'
    },
  ].map(({ entries, location, dateRange, optionType }, index) =>
    it(`should return 'computedGraphData' (index: ${index})`, () => {
      expect(graphCountManager.computedGraphData(List(TestEntry)(entries), location, dateRange, optionType).toJS()).toMatchSnapshot();
    })
  );


});
