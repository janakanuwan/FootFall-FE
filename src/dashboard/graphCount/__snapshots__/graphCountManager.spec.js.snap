// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphCountManager should return 'graphData' (index: 0) 1`] = `
Array [
  Object {
    "IN": 0,
    "NAME": "1:00 AM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 1,
    "NAME": "2:00 AM",
    "OUT": 0,
    "PRESENCE": 1,
  },
  Object {
    "IN": 3,
    "NAME": "3:00 AM",
    "OUT": 1,
    "PRESENCE": 2,
  },
  Object {
    "IN": 2,
    "NAME": "4:00 AM",
    "OUT": 3,
    "PRESENCE": -1,
  },
  Object {
    "IN": 5,
    "NAME": "5:00 AM",
    "OUT": 6,
    "PRESENCE": -1,
  },
  Object {
    "IN": 4,
    "NAME": "6:00 AM",
    "OUT": 2,
    "PRESENCE": 2,
  },
  Object {
    "IN": 3,
    "NAME": "7:00 AM",
    "OUT": 0,
    "PRESENCE": 3,
  },
  Object {
    "IN": 0,
    "NAME": "8:00 AM",
    "OUT": 1,
    "PRESENCE": -1,
  },
  Object {
    "IN": 0,
    "NAME": "9:00 AM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "10:00 AM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "11:00 AM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "12:00 PM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "1:00 PM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "2:00 PM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "3:00 PM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "4:00 PM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "5:00 PM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "6:00 PM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "7:00 PM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "8:00 PM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "9:00 PM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "10:00 PM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "11:00 PM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "11:59 PM",
    "OUT": 0,
    "PRESENCE": 0,
  },
]
`;

exports[`graphCountManager should return 'graphData' (index: 1) 1`] = `
Array [
  Object {
    "IN": 0,
    "NAME": "1:00 AM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 1,
    "NAME": "2:00 AM",
    "OUT": 0,
    "PRESENCE": 1,
  },
  Object {
    "IN": 3,
    "NAME": "3:00 AM",
    "OUT": 1,
    "PRESENCE": 2,
  },
  Object {
    "IN": 2,
    "NAME": "4:00 AM",
    "OUT": 3,
    "PRESENCE": -1,
  },
  Object {
    "IN": 5,
    "NAME": "5:00 AM",
    "OUT": 6,
    "PRESENCE": -1,
  },
  Object {
    "IN": 4,
    "NAME": "6:00 AM",
    "OUT": 2,
    "PRESENCE": 2,
  },
  Object {
    "IN": 3,
    "NAME": "7:00 AM",
    "OUT": 0,
    "PRESENCE": 3,
  },
  Object {
    "IN": 0,
    "NAME": "8:00 AM",
    "OUT": 1,
    "PRESENCE": -1,
  },
  Object {
    "IN": 0,
    "NAME": "9:00 AM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "10:00 AM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "11:00 AM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "12:00 PM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "1:00 PM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "2:00 PM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "3:00 PM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "4:00 PM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "5:00 PM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "6:00 PM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "7:00 PM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "8:00 PM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "9:00 PM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "10:00 PM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "11:00 PM",
    "OUT": 0,
    "PRESENCE": 0,
  },
  Object {
    "IN": 0,
    "NAME": "11:59 PM",
    "OUT": 0,
    "PRESENCE": 0,
  },
]
`;

exports[`graphCountManager should return 'graphData' (index: 2) 1`] = `
Array [
  Object {
    "IN": 4,
    "NAME": "May-01",
    "OUT": 3,
    "PRESENCE": 1,
  },
  Object {
    "IN": 2,
    "NAME": "May-02",
    "OUT": 1,
    "PRESENCE": 1,
  },
  Object {
    "IN": 3,
    "NAME": "May-03",
    "OUT": 1,
    "PRESENCE": 2,
  },
  Object {
    "IN": 2,
    "NAME": "May-04",
    "OUT": 3,
    "PRESENCE": -1,
  },
]
`;

exports[`graphCountManager should return 'graphData' (index: 3) 1`] = `
Array [
  Object {
    "IN": 6,
    "NAME": "May-01:May-07",
    "OUT": 4,
    "PRESENCE": 2,
  },
  Object {
    "IN": 3,
    "NAME": "May-08:May-14",
    "OUT": 1,
    "PRESENCE": 2,
  },
  Object {
    "IN": 8,
    "NAME": "May-15:May-18",
    "OUT": 8,
    "PRESENCE": 0,
  },
]
`;

exports[`graphCountManager should return 'graphData' (index: 4) 1`] = `
Array [
  Object {
    "IN": 6,
    "NAME": "May 2018",
    "OUT": 4,
    "PRESENCE": 2,
  },
  Object {
    "IN": 3,
    "NAME": "Jun 2018",
    "OUT": 1,
    "PRESENCE": 2,
  },
  Object {
    "IN": 2,
    "NAME": "Jul 2018",
    "OUT": 3,
    "PRESENCE": -1,
  },
  Object {
    "IN": 10,
    "NAME": "Aug 2018",
    "OUT": 7,
    "PRESENCE": 3,
  },
  Object {
    "IN": 0,
    "NAME": "Sep 2018",
    "OUT": 0,
    "PRESENCE": 0,
  },
]
`;
